File: pcap/pcap_monitor.py
#####
#Create a Python script that starts overlapping  `tcpdump` processes for a given duration and then 
# triggers the `pcap_modbus.py` or `pcap+dnp3.py` programs.
# The captures are each for a set period of time and have an 'overlap' period to ensure continuous data 
# capture
#The `subprocess` module is used for process management and `time` for delays. 
# python pcap_monitor.py config_file (pcap_monitor.json ) log_dir (/var/log/gcom)  # Replace with actual interface names
# uses a config file default name pcap_monitor.json with this data
# note this file will be produced by the pcap_fil_scanner.py program
#{
#    "interfaces": ["eth0", "wlan0"]
#}

####

import os
import json
import argparse
import signal
import subprocess
import time
from datetime import datetime
dt_format = "%Y%m%d_%H%M%S"

procs = {}


def ensure_log_dir(log_dir):
    # Create log directory if it does not exist
    if not os.path.exists(log_dir):
        try:
            os.makedirs(log_dir)
        except OSError as e:
            print(f"Error: {e.strerror} - Log directory ({log_dir}) could not be created.")
            exit(1)

def run_pcap_file_scanner():
    scanner_script = '/usr/local/bin/pcap_file_scanner.py'  # Update with the actual path
    try:
        # Run the scanner script and wait for it to complete
        subprocess.run(['python3', scanner_script], check=True)
    except subprocess.CalledProcessError as e:
        print(f"Failed to run pcap_file_scanner: {e}")
        # Handle error or possibly exit if the scanner is critical to subsequent operations

# Ensure the log directory exists at the beginning of your script
#ensure_log_dir(args.logdir)

def signal_handler(signum, frame):
    print(f"Received termination signal {signum}")
    # Terminate all tcpdump processes
    for proc in procs.values():
        proc.terminate()
    print("All captures terminated. Exiting.")
    exit(0)  # Exit the program

def load_config(config_file):
    with open(config_file, 'r') as file:
        config = json.load(file)
    return config

def start_tcpdump(interface, log_dir, duration=300):
    # Generate a unique filename for each pcap
    timestamp = datetime.now().strftime(dt_format)
    filename = f"{log_dir}/capture_{interface}_{timestamp}.pcap"

    # Start tcpdump
    tcpdump_cmd = ["tcpdump", "-i", interface, "-w", filename]
    proc = subprocess.Popen(tcpdump_cmd)

    # Return the process and filename for later use
    return proc, filename

def manage_captures(interfaces, log_dir, overlap=10, duration=300):
    global procs
    try:
        while True:
            filenames = {}
            
            # Start captures for each interface
            for interface in interfaces:
                procs[interface], filenames[interface] = start_tcpdump(interface, log_dir, duration)

            # Wait for duration minus overlap
            time.sleep(duration - overlap)

            # Stop previous captures and analyze them
            for interface in interfaces:
                procs[interface].terminate()
                subprocess.run(["python3", "pcap_modbus.py", filenames[interface]])

                # Immediately start the next capture for no gap
                procs[interface], filenames[interface] = start_tcpdump(interface, duration)

            # Wait for the overlap period
            time.sleep(overlap)

    except KeyboardInterrupt:
        # On a keyboard interrupt/termination signal, terminate all tcpdump processes
        print("\nTerminating all captures...")
        for proc in procs.values():
            proc.terminate()
        print("All captures terminated.")

# Main execution
if __name__ == "__main__":
    # Setup argument parser
    parser = argparse.ArgumentParser(description="Continuously capture network packets on multiple interfaces")
    parser.add_argument("config_file", help="Configuration file containing the list of interfaces", nargs='?', default="pcap_monitor.json")
    parser.add_argument("--logdir", help="Directory to store log files", default="/var/log/gcom")

    args = parser.parse_args()

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    ensure_log_dir(args.logdir)

    run_pcap_file_scanner()

    try:
        config = load_config(args.config_file)
        interfaces = config.get('interfaces', [])  # Provide default empty list if not found

        # Check if interfaces are provided
        if interfaces:
            manage_captures(interfaces,args.logdir)
        else:
            print("No interfaces found in configuration file.")
    except FileNotFoundError:
        print(f"Configuration file {args.config_file} not found.")
    except json.JSONDecodeError:
        print(f"Error decoding JSON from the file {args.config_file}.")

File: pcap/go/pcap_monitor.go
package main

import (
	//"encoding/json"
	"flag"
	"fmt"
	//"io/ioutil"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"syscall"
	"time"
	"strings"
    // "bytes"
    // "encoding/binary"
)

type Config struct {
	Interfaces []string `json:"interfaces"`
}

func main() {
	//configFile := flag.String("c", "/var/log/gcom/pcap_monitor.json", "Configuration file containing the list of interfaces")
    logDir := flag.String("l", "/var/log/gcom", "Directory to store log files")
    logTime := flag.Int("t", 300, "Time for each file log files")

	// Define a custom help flag
	//help := flag.Bool("h", false, "Display this help message")
    // Rest of your main function...
	ifList := flag.String("i", "", "Comma-separated list of interfaces")

	flag.Parse()

	// if *help {
	// 	fmt.Printf("Usage of %s:\n", os.Args[0])
	// 	flag.PrintDefaults()
	// 	return
	// }
	
	// Validate arguments
    if *ifList == "" {
        flag.Usage()
		fmt.Printf(" For example :./pcap_monotor -i eth0,eth1  \n")
		fmt.Printf("              ./pcap_monotor -i eth0,eth1  -t 20 -l /var/logs/mydir \n")
        os.Exit(1)
    }

    //fmt.Printf("Config file: %s\n", *configFile)
    fmt.Printf("Log directory: %s\n", *logDir)
	//configFile := flag.String("config_file", "/var/log/gcom/pcap_monitor.json", "Configuration file containing the list of interfaces")
	//logDir = flag.String("logdir", "/var/log/gcom", "Directory to store log files")

	hostname, err := os.Hostname()
	if err != nil {
		fmt.Printf("Failed to get hostname: %v\n", err)
		return
	}

	fmt.Printf("Hostname: %s\n", hostname)
	
	if *ifList == "" {
		fmt.Printf("please provide an interface list\n")
		return
	}
	ifStrs := strings.Split(*ifList, ",")
	// config, err := loadConfig(*configFile)
	// if err != nil {
	// 	fmt.Printf("Failed to load config: %v\n", err)
	// 	return
	// }

	if len(ifStrs) == 0 {
		fmt.Println("No interfaces found on command line.")
		return
	}

	// Create log directory if it does not exist
	err = os.MkdirAll(*logDir, 0755)
	if err != nil {
		fmt.Printf("Failed to create log directory: %v\n", err)
		return
	}

	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)

	go func() {
		<-c
		fmt.Println("\nReceived termination signal. Exiting.")
		os.Exit(0)
	}()

	manageCaptures(hostname, ifStrs, *logDir, *logTime)
}

// func loadConfig(filePath string) (Config, error) {
// 	var config Config
// 	data, err := ioutil.ReadFile(filePath)
// 	if err != nil {
// 		return config, err
// 	}
// 	err = json.Unmarshal(data, &config)
// 	return config, err
// }

func manageCaptures(hostname string, interfaces []string, logDir string, logTime int) {
	overlap := 10 * time.Second
	duration := time.Duration(logTime * 60)

	for _, intf := range interfaces {
		if intf == "lo" {
			continue
		}
		go startTcpdump(hostname, intf, logDir, duration, overlap)
	}

	select {}
}

func startTcpdump(hostname, intf, logDir string, duration, overlap time.Duration) {
	for {
		timestamp := time.Now().Format("20060102_150405")
		hostname = "mything"
		logDir = "/tmp/pcap"
		pcapFilename := filepath.Join(logDir, fmt.Sprintf("%s_%s_%s.pcap", hostname, intf, timestamp))
		logFilename := filepath.Join(logDir, fmt.Sprintf("%s_%s_%s.log", hostname, intf, timestamp))

		fmt.Printf("Starting tcpdump on %s, output file: %s\n", intf, pcapFilename)
		//cmd := exec.Command("timeout", "60", "tcpdump", "-i", intf, "-w", pcapFilename)
		cmd := exec.Command("sh", "run_tcp.sh", "60", intf, pcapFilename)
		logFile, err := os.Create(logFilename)
		if err != nil {
			fmt.Printf("Failed to create log file: %v\n", err)
			continue
		}
		cmd.Stdout = logFile
		cmd.Stderr = logFile

		if err := cmd.Start(); err != nil {
			fmt.Printf("Failed to start tcpdump: %v\n", err)
			logFile.Close()
			continue
		}

		time.Sleep(duration - overlap)

		if err := cmd.Process.Signal(syscall.SIGINT); err != nil {
			fmt.Printf("Failed to send SIGINT to tcpdump: %v\n", err)
		}

		// Wait for tcpdump to finish
		cmd.Wait()
		logFile.Close()

		// You can add code here to call pcap_modbus or pcap_dnp3 if needed

		time.Sleep(overlap)
	}
}